### PR TITLE
Enable TargetFrameworkName attribute

### LINKED ISSUE #26456: .dll in Microsoft.NET.Build.Extensions\net461 targets .NET Framework 4.6.1
_From @jairbubbles on September 27, 2017 10:4_

I just noticed that most .dll in `C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Microsoft\Microsoft.NET.Build.Extensions\net461\` directory target .NET Framework 4.7 which does not seem right.

Here is a dotPeek screenshot:
image

_Copied from original issue: dotnet/sdk#1611_

### LINKED ISSUE #26457: CoreFx assemblies lack the TargetFrameworkAttribute
I notice this when looking at https://github.com/dotnet/corefx/issues/26456.  It appears to still be the case.  Filing here, though the right fix may be in buildtools.

### PR DESCRIPTION
**Pull request by ViktorHofer**
Remove the opt-out of the TargetFrameworkName attribute generation and
fix the existing test to actually test the custom attribute being set
and read correctly by the AppContext class.

Fixes #26456
Fixes #26457

See previous attempt https://github.com/dotnet/corefx/pull/32680 for discussion.

### CODE REVIEW

File: src/System.Runtime.Extensions/tests/CustomAttributesTestApp/Program.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Reflection;
+using System.Threading;
+
+namespace CustomAttributesTestApp
```
Comment (jkotas) [inline src/System.Runtime.Extensions/tests/CustomAttributesTestApp/Program.cs]: Nit: A better name for this may be `TargetFrameworkNameTestApp`

File: src/System.Runtime.Extensions/tests/System/AppDomainTests.cs [hunk 2]
```diff
@@ -41,17 +41,20 @@ public void RelativeSearchPath_Is_Null()
         }
 
         [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap, "Process.Start not allowed inside AppContainer")]
```
Comment (jkotas) [inline src/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: I believe that `Process.Start` works in AppContainer just fine. Is the actual problem the limitations of the CoreFX test infrastructure?

Comment (ViktorHofer) [reply src/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: I did not try running the test on UWP as I believed that from inside an AppContainer you can't start another process.

File: src/System.Runtime.Extensions/tests/CustomAttributesTestApp/Program.cs (hunk 1 above)
Comment (ViktorHofer) [reply src/System.Runtime.Extensions/tests/CustomAttributesTestApp/Program.cs]: I used that name as in future we might want to test other attributes here as well.

File: src/System.Runtime.Extensions/tests/System/AppDomainTests.cs (hunk 2 above)
Comment (ViktorHofer) [reply src/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: I'll try it out.

File: src/System.Runtime.Extensions/tests/CustomAttributesTestApp/CustomAttributesTestApp.csproj [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    <Configurations>netcoreapp-Debug;netcoreapp-Release</Configurations>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="Program.cs" />
+    <Compile Include="AssemblyAttributes.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="$(CommonPath)\System\Diagnostics\CodeAnalysis\ExcludeFromCodeCoverageAssemblyAttribute.cs">
+      <Link>Common\System\Diagnostics\CodeAnalysis\ExcludeFromCodeCoverageAssemblyAttribute.cs</Link>
+    </Compile>
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/tests/CustomAttributesTestApp/CustomAttributesTestApp.csproj]: Why is this needed?

File: src/System.Runtime.Extensions/tests/CustomAttributesTestApp/Program.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Runtime.Extensions/tests/CustomAttributesTestApp/Program.cs]: FWIW, I prefer @jkotas' naming suggestion.  If in the future we want to expand it to encompass other things, it can be renamed then as desired.

File: src/System.Runtime.Extensions/tests/System/AppDomainTests.cs [hunk 4]
```diff
@@ -41,17 +41,20 @@ public void RelativeSearchPath_Is_Null()
         }
 
         [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap, "Corefx has limitations to build a UWP executable that can be launched directly using Process.Start")]
         public void TargetFrameworkTest()
         {
-            string targetFrameworkName = ".NETCoreApp,Version=v2.1";
-            if (PlatformDetection.IsInAppContainer)
+            int expectedExitCode = 0;
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: const int ExpectedExitCode

File: src/System.Runtime.Extensions/tests/CustomAttributesTestApp/AssemblyAttributes.cs [hunk 5]
```diff
@@ -0,0 +1,6 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+[assembly: System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverageAssembly]
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/tests/CustomAttributesTestApp/AssemblyAttributes.cs]: Why do we need this?

File: src/System.Runtime.Extensions/tests/CustomAttributesTestApp/CustomAttributesTestApp.csproj (hunk 3 above)
Comment (ViktorHofer) [reply src/System.Runtime.Extensions/tests/CustomAttributesTestApp/CustomAttributesTestApp.csproj]: Opened https://github.com/dotnet/corefx/pull/41674 in response to this.

**Comment by ViktorHofer** [comment]
@stephentoub can you please review again? Addressed your feedback.

File: src/System.Runtime.Extensions/tests/System/AppDomainTests.cs [hunk 6]
```diff
@@ -41,17 +41,20 @@ public void RelativeSearchPath_Is_Null()
...
         {
-            string targetFrameworkName = ".NETCoreApp,Version=v2.1";
-            if (PlatformDetection.IsInAppContainer)
+            const int ExpectedExitCode = 0;
+            const string AppName = "TargetFrameworkNameTestApp.dll";
+            var psi = new ProcessStartInfo();
+            psi.FileName = RemoteExecutor.HostRunner;
+            psi.Arguments = $"{AppName} {ExpectedExitCode}";
+
+            using (Process p = Process.Start(psi))
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: Nit: you don't need the psi... you can just pass the name and args directly to start

**Review by stephentoub — APPROVED** [review]

File: src/System.Runtime.Extensions/tests/System/AppDomainTests.cs (hunk 6 above)
Comment (ViktorHofer) [reply src/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: I'll leave this as is for now as we have multiple places where we are following this pattern.

**Comment by ViktorHofer** [comment]
Thanks for the review!

File: src/System.Runtime.Extensions/tests/System/AppDomainTests.cs (hunk 6 above)
Comment (stephentoub) [reply src/System.Runtime.Extensions/tests/System/AppDomainTests.cs]: For reference can you link to some?

Comment (ViktorHofer) [reply src/System.Runtime.Extensions/tests/System/AppDomainTests.cs:53]: Sorry, I never followed on this and just saw it. Couldn't remember which places I meant back then.
